service/storage_service.cc

```diff
@@ -5678,6 +5678,28 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
 
         switch (cmd.cmd) {
             case raft_topology_cmd::command::barrier:
+                // This barrier might have been issued by the topology coordinator
+                // as a step in enabling a feature, i.e. it noticed that all
+                // nodes support some feature, then issue the barrier to make
+                // sure that all nodes observed this fact in their local state
+                // (a node cannot revoke support for a feature after that), and
+                // after receiving a confirmation from all nodes it will mark
+                // the feature as enabled.
+                //
+                // However, it might happen that the node handles this request
+                // early in the boot process, before it did the second feature
+                // check that happens when the node updates its metadata
+                // in `system.topology`. The node might have committed a command
+                // that advertises support for a feature as the last node
+                // to do so, crashed and now it doesn't support it. This should
+                // be rare, but it can happen and we can detect it right here.
+                if (!_topology_updated_with_local_metadata) {
+                    // TODO: Somehow make it stop the boot process if it fails
+                    const auto& enabled_features = _topology_state_machine._topology.features.enabled_features;
+                    const auto unsafe_to_disable_features = _topology_state_machine._topology.features.calculate_not_yet_enabled_features();
+                    _feature_service.check_features(enabled_features, unsafe_to_disable_features);
```

Comment:
It should be enough to check that there is no feature in  `_topology_state_machine._topology.features.enabled_features` that are also not in  `_feature_service.supported_feature_set()`. We can also do it unconditionally. This is not a fast path.

Reply:
I agree with what you wrote, but instead of `enabled_features` we should use `unsafe_to_disable_features`.

Reply:
Actually, I'd like to stick to `check_features` due to the message it produces via exception - some tests expect this message to be printed into logs.

Reply:
I would like to get rid of `_topology_updated_with_local_metadata`. Simple check can be done unconditionally. I wouldn't want tests dictate the code. Can we print the same message if the check fails?

If we cannot get rid of `_topology_updated_with_local_metadata` then it should not be a bool but a feature that is waited here instead of the check.

Reply:
I removed `_topology_updated_with_local_metadata`.

I agree that the tests should not dictate the code, but I'd also like to keep the error messages to be the same for the sake of users reading the logs (imagine that you get a different message 1% of the times - weird and confusing). The exception message is also slightly different depending on the circumstances, so I resorted to catching the exception and printing its message in the logs.

Reply:
I actually usually prefer to have slightly different messages. I know what code path failed then. It can be just a message prefix like the function name.

Reply:
Hm, alright, actually the message is now slightly different that what is printed in the usual case - it will be prefixed with `raft topology: feature check during barrier failed: `, but the part that follows (extracted from the exception) is the same.

Reply:
Perfect.